Fix for Root Dockerfile building from Dockerhub not source and also adjusting Makefile

nilshee opened the pull request:
Idk if there was some configuration error on my side. 
But for me it looked like dev/docker-compose.full.yml was using the Dockerfiles from the root directory because of which it pulled directly from Dockerhub which is kinda useless in dev.
only fixes docker-compose -f docker-compose.full.yml up/build

deleted whole dev/dockerfile.dev (which was not working at all) and copied stuff over from dockerfile.base and dockerfile

nilshee commented:
Just saw that @jemand771 already seems to be working on that in a more professional way in #17 . But nevertheless this is a Quickifix for manual builds

jemand771 commented:
Hi there! :) we're still trying to figure out what all the different build jobs / configurations do, so let's make sure I understand this correctly:
* the `dev/docker-compos-full.yml` stack used to build the main `Dockerfile`
* the main `Dockerfile` currently pulls an image from DockerHub instead of multi-stage building the `:base` image (what i'm partially working on)
* you've modified the `dev/Dockerfile` to do what `./Dockerfile` and `./Dockerfile.base` would do if put together
* the compose stack now uses said modified dockerfile instead of the "normal" one in the repo root while still using the normal build context

is that correct so far?
Questions I might have right away (don't worry if you don't know, I'll look into it myself at some point) and things to do:
- [x] me: test your dockerfile / the whole stack
- [x] you/me: check whether any other stacks, github workflows, and make targets were using the dev dockerfile (that you modified) before -> maybe adjust those aswell if needed
- [x] make sure that nothing else breaks

sorry if this seems somewhat unprofessional, as I've said we're still figuring things out ^^

**edit:** this comment is somewhat outdated after the force push

Commit:
Dockerfile build local files not remote files

Commit:
reduce size of Image

Commit:
remove base-image

Commit:
adjust image names to be statping-ng/statping-ng

nilshee commented:
This correct so far. But I broke some things.

I corrected those and changed my pull request to change the current Root Dockerfile (I took some inspiration from you. Now it builds the dev/docker-compose.full.yml correctly). I guess the base image would also work but personally I would remove the whole base image I don't understand why we need it. It only builds the statping binary to be copied to another img.

jemand771 commented on Dockerfile:
```diff
@@ -1,12 +1,55 @@
+# /go/bin/statping - statping binary
+# /root/sassc/bin/sassc - sass binary
+# /statping - Vue frontend (from frontend)
+
```
nitpick: I'd keep this comment (I think you accidentally deleted it)
```suggestion

# Statping main Docker image that contains all required libraries
```

jemand771 commented on Dockerfile:
```diff
@@ -1,12 +1,55 @@
-COPY --from=base /go/bin/statping /usr/local/bin/
-COPY --from=base /root/sassc/bin/sassc /usr/local/bin/
-COPY --from=base /usr/local/share/ca-certificates /usr/local/share/
+COPY --from=backend /go/bin/statping /usr/local/bin/
+COPY --from=backend /root/sassc/bin/sassc /usr/local/bin/
```
The original Dockerfile also copied the ca certificates from the backend stage. I don't know if that's necessary, but I'd keep this line for now just in case
```suggestion
COPY --from=backend /go/bin/statping /usr/local/bin/
COPY --from=backend /root/sassc/bin/sassc /usr/local/bin/
COPY --from=backend /usr/local/share/ca-certificates /usr/local/share/
```

nilshee replied:
well we can do that, but ca-certificates is already installed by `RUN apk --no-cache add libgcc libstdc++ ca-certificates curl jq && update-ca-certificates`

jemand771 replied:
Good point - we'll take another look at that at some point in the future

jemand771 commented on Dockerfile:
```diff
@@ -20,4 +63,4 @@ EXPOSE $PORT
 HEALTHCHECK --interval=60s --timeout=10s --retries=3 CMD curl -s "http://localhost:$PORT/health" | jq -r -e ".online==true"
 
-CMD statping --port $PORT
+CMD statping --port $PORT
```
nitpick: newline at end of file (generally good practice)
```suggestion
CMD statping --port $PORT

```

jemand771 commented on Makefile:
```diff
@@ -126,13 +123,6 @@ docker-vue:
-push-base: clean compile docker-base
-	docker push statping-ng/statping-ng:base
-
-push-vue: clean compile docker-base docker-vue
-	docker push statping-ng/statping-ng:base
-	docker push statping-ng/statping-ng:vue
-
```
you removed `push-vue` (which I don't think is used by anything) but kept `docker-vue`. I think we should remove either both or none of them

nilshee replied:
I would remove both, but that's mainly because I don't understand what to do with them.

nilshee replied:
removed in new commit

jemand771 commented on Makefile:
```diff
@@ -285,16 +275,9 @@ download-key:
 	wget -O statping.gpg $(SIGN_URL)
 	gpg --import statping.gpg
 
-# push the :dev docker tag using curl
-dockerhub-dev:
-	docker build --build-arg VERSION=${VERSION} -t statping-ng/statping-ng:dev --no-cache -f Dockerfile.base .
-	docker push statping-ng/statping-ng:dev
-
```
Note: (partially to myself because I _will_ forget this)
removing the push target for `:dev` while keeping the build target is a good thing since we don't want to push the dev tag but keep it for local testing (avoid confusion between a local-only tag and one that was pulled from somewhere)

you can just click "resolve conversation" on this comment once you've seen it ^^

nilshee replied:
I mostly removed it because it used the base image which is weird. (Why push such a large image to Dockerhub? )

jemand771 commented on docker-compose.yml:
```diff
@@ -3,7 +3,7 @@ version: '2.3'
   statping:
     container_name: statping
-    image: aboutcher/statping-ng:dev
+    image: statping-ng/statping-ng:dev
```
that's not right, but it's not your fault ;) https://hub.docker.com/r/statping-ng/statping-ng doesn't exist, but we should probably switch to that at some point ^^
```suggestion
    image: aboutcher/statping-ng:dev
```

nilshee replied:
that's where the problems start :D by using aboutcher/statping the image will always be pulled directly from dockerhub. or we need to build every image as aboutcher/statping locally. But for now we can keep it. I guess

jemand771 replied:
ooh I see. for now, there's kind of a "split workflow":
* if you want to deploy statping somewhere, you use the aboutcher image -> pulled from docker hub
* if you want to work on statping, you use the statping-ng image in combination with the compose files in dev/ -> built locally

jemand771 reviewed (changes requested):
there's some small things left that I'd like to change and/or talk about
(don't be scared by the big red **changes requested** banner and correct me if I'm wrong with anything :) )

also note that GitHub's suggestion view doesn't render all diffs correctly, you'll have to use your imagination on the ones that look weird ^^

nilshee committed:
add description for Image with binary

Co-authored-by: Willy <jemand771@gmx.net>

nilshee committed:
Apply suggestions from code review

Co-authored-by: Willy <jemand771@gmx.net>

nilshee committed:
remove docker-vue

jemand771 reviewed (approved):
lgtm so far :)

jemand771 commented:
@adamboutcher you mind taking a look at this? in case I haven't missed anything, this is ready to merge

also, we'll need to think about the docker image names and tags again (unstable vs stable, :dev tag, and so on)

adamboutcher commented:
I'll try and review this later today, worst case it'll be tomorrow.

adamboutcher commented:
Only issues I can see is that the images push to statping-ng/statping-ng but we don't have a dockerhub org to do that, so that's why is its under my name, unless we do have that via someone?

nilshee commented:
That means we have to change every statping-ng/statping-ng to aboutcher/statping-ng to fit to your dockerhub acc.

adamboutcher commented:
> That means we have to change every statping-ng/statping-ng to aboutcher/statping-ng to fit to your dockerhub acc. 

adamboutcher/statping-ng

But yes, unless we can get an organisational account on dockerhub, which I believe costs.

Commit:
fixed naming for dockerhub

nilshee commented:
fixed it for `make dockerhub`

Commit:
fix naming in docker-compose

jemand771 commented:
i mean we _could_ just create a personal account `statping-ng` using some alt email - not sure if we want to do that

adamboutcher commented:
> i mean we _could_ just create a personal account `statping-ng` using some alt email - not sure if we want to do that

That could be viable, I have an email address at my domain already setup for somethings so maybe I should do that on dockerhub.

So I've signed up with my statping email, however you can't have statping-ng as the username as that's unsupported 😔 so I've chosen statpingng. I however can't verify it yet as I haven't recieved the email 😔.

adamboutcher commented:
So I'm not sure about the docker repo change, we can either stick with adamboutcher/statping-ng (already has a few thousand pulls) or move to statpingng/statping-ng.

However snaps (once they work) will retain my name, so maybe it's work keeping it under my own purely for some basic consistency, if we could have have statping-ng in dockerhub then I'd be all over it.

jemand771 commented:
i agree. `statpingng` is somewhat ugly.
one thing to note: i'm planning on releasing to ghcr.io as part of #17 (if that works the way I want it to), so that'd become `ghcr.io/statping-ng/statping-ng`. having two different image names would somewhat suck but statping-ng + adamboutcher is better than statpingng + statping-ng

adamboutcher commented:
> i agree. `statpingng` is somewhat ugly.
> one thing to note: i'm planning on releasing to ghcr.io as part of #17 (if that works the way I want it to), so that'd become `ghcr.io/statping-ng/statping-ng`. having two different image names would somewhat suck but statping-ng + adamboutcher is better than statpingng + statping-ng

We also have the quay.io repo quay.io/statping-ng/statping-ng which supposedly autobuilds, so maybe we keep adamboutcher/statping-ng as a fall back but push quay or ghcr as our main container repo?

nilshee commented:
so should we change `./docker-compose.yml` to the quay.io images? And neglect Dockerhub?
Because `statpingmg/statping-ng` is really ugly.
In my [commit ](https://github.com/statping-ng/statping-ng/pull/25/commits/a222b0148a00ebe8b69dc30fb0b94e90ad9310ea) I renamed everthing which relates to dockerhub to adamboutcher/statping-ng and everything which is used for local development to `statping-ng/statping-ng`. (So no risk pushing it accidently 😉 )

adamboutcher commented:
Personally we should pick one as our main but still provide on the others, maybe it's worth setting the ghcr one as the main especially if @jemand771 has written some nice magic scripts that work with it.

adamboutcher commented:
Entirely upto @jemand771 with this one, unless someone else has a better suggestion.

jemand771 commented:
the "magic script" should be able to accept any docker repo as a target so that doesn't really matter.
ghcr has the advantage that it's right here on the same page as the source code

adamboutcher commented:
So do you think we should use ghcr then?

Also does your script allow us to push to dockerhub as well as ghcr?

Finally, are you ready to merge your changes and these suggested by @nils1323 too?

jemand771 commented:
1: ghcr sounds good
2: docker-metadata-action (I think that's what it's called) can take a list of image names, so pushing the same image to multiple repos is really simple
3: #17 still needs a lot of work since I only made a todo list and didn't change much by now. I won't get a lot done in the next two weeks (exams next week, vacation the following week). This PR looks ready. There's still `statping-ng/statping-ng` in some places but that doesn't really matter

I think in the future we can/should remove (or at least reduce) the docker-related make targets since you don't need anything other than `docker build -t <some tag> .`. #17 probably won't use the makefile for docker releases anyway.
I don't think we'll do any manual docker pushes --> the image/tag used locally doesn't matter

Except for the image names (which don't really matter for local builds as long as they're consistent) this looks ready to merge. If actions break somehow I'll fix them in #17 (there currently aren't any new code changes to release anyway)

adamboutcher commented:
So we can merge this and deal with your changes later?

nilshee commented:
Currently I have the problem that I can not build changes to the frontend. I can't find the reason, it somehow tries to access GitHub.com/statping instated of GitHub/statping-ng and somehow bricks the whole frontend. It doesn't reflect any changes to the .Vue files.
The telegram part ist bricked cause you can only supply strings (da🦆), and so u can not supply chat IDS (which is way more useful than supplying any Public @channel ).

jemand771 commented:
@nils1323 are your build issues related to this PR? when I `docker build .` this, I get the correct image (e.g. statping-ng banner etc.)
some things (like the docker-compose.full.yml) are still broken, but the important things you changed seem to work (./Dockerfile and ./docker-compose.yml)

nilshee commented:
> @nils1323 are your build issues related to this PR? when I `docker build .` this, I get the correct image (e.g. statping-ng banner etc.)
> some things (like the docker-compose.full.yml) are still broken, but the important things you changed seem to work (./Dockerfile and ./docker-compose.yml)

True, it is somewhat working again. Idk it could be that i destroyed my local repository. I cloned my remote into a new dir and now it is working again. This changes can be merged.

adamboutcher commented:
I'll merge this now then.
